Rust duration type

adespawn opened the pull request:
This is an implementation of one of the datatype used by the database. The current approch keeps the internal data of the type in object created in rust, while all the processing of this object is done in JS part (it usses mostly unchanged datastax code).  Appart from diffrent behaviour in case of providing unexpected type in the constructor, everything should work the same way as in old implementation. If not, treat it as a bug.

This is a example implementation. The goal is to have some kind of example how to implement datastax custom types.
Current approach keeps the data in the rust object, but leaves all the processing that is done in the js part unchanged.
As this type is analog to ``CqlDuration`` used by the rust driver, there should be a way to create ``Duration`` object from ``CqlDuration``, which is implemented. 

Both comments about this specific implementation, and general approach are welcomed.

Tests expected to pass:
[There is new, temporary workflow with just the tests expected to pass]
``test/unit/duration-type-tests.js``

adespawn commented on lib/types/duration.js:
```diff
@@ -406,7 +473,7 @@ Builder.prototype.addMillis = function(millis) {
  * @param {String|Long} micros
  * @return {Builder}
  */
-Builder.prototype.addMicros = function(micros) {
+Builder.prototype.addMicros = function (micros) {
   const value = typeof micros === 'string' ? Long.fromString(micros) : micros;
   this._validateOrder(9);
```
@Stapox35 Added JS formatter in #30.

wprzytula commented on lib/types/duration.js:
```diff
@@ -102,15 +96,15 @@ Duration.prototype.toString = function () {
     return dividend.modulo(divisor);
   }
-  if (this.months < 0 || this.days < 0 || this.nanoseconds.isNegative()) {
+  if (this.internal.months < 0 || this.internal.days < 0 || nanoseconds.isNegative()) {
```
why did `this.nanoseconds` change into `nanoseconds`? Are both equivalent in JS?

adespawn replied:
this.internal.nanoseconds refers to the value in rust object. In JS it's of bigint type. This datastax code treats nanoseconds as Long, so we need to convert the type, which we keep in local variable nanoseconds.

wprzytula replied:
My bad, I didn't notice the variable. Nvm.

wprzytula commented on lib/types/duration.js:
```diff
@@ -102,15 +96,15 @@ Duration.prototype.toString = function () {
 
-  if (!this.nanoseconds.equals(Long.ZERO)) {
-    const nanos = this.nanoseconds.isNegative() ? this.nanoseconds.negate() : this.nanoseconds;
+  if (!nanoseconds.equals(Long.ZERO)) {
```
ditto

wprzytula commented on lib/types/duration.js:
```diff
@@ -127,6 +121,8 @@ Duration.prototype.toString = function () {
  * @returns {Duration}
  */
 Duration.fromBuffer = function (buffer) {
+  if (!buffer instanceof Buffer)
+    throw new TypeError('Expected Buffer type');
```
Is the goal of this an earlier incompatible type detection?

adespawn replied:
Sort of. But as this function is unchanged from original, it may make more sense to don't do it.

wprzytula commented on src/result.rs:
```diff
@@ -181,6 +183,14 @@ impl CqlValueWrapper {
     }
   }
 
+  #[napi]
+  pub fn get_duration(&self) -> napi::Result<DurationWrapper> {
+    match self.internal.as_cql_duration() {
+      Some(r) => Ok(DurationWrapper::from_cql_duration(r)),
+      None => Err(Error::new(Status::GenericFailure, "Error")),
+    }
+  }
```
Of course the error message must be more informative. Maybe `Error: the underlying CQL value is not of Duration type`?

wprzytula replied:
Also, please check which `Status` variant is the best suit here. Perhaps it's `GenericFailure` indeed, but let's make sure.

adespawn replied:
Here is list of all possible ``Status`` values: [code](https://github.com/napi-rs/napi-rs/blob/de3a60dfc1e59b3a0a493cf8d955506f7ab39d86/crates/napi/src/status.rs#L7-L35) I can't find any appropriate documentation, but it seems all other types are for receiving invalid type of argument.

wprzytula commented on src/types/duration.rs:
```diff
@@ -0,0 +1,59 @@
+  pub fn new(months: i32, days: i32, ns_bigint: BigInt) -> napi::Result<Self> {
+    let ns = ns_bigint.get_i64();
+    if !ns.1 {
+      return Err(Error::new(
+        Status::GenericFailure,
+        "Invalid use: Should not happen?",
+      ));
+    }
```
What can trigger this error? Please reflect the possible cause in the error message.

wprzytula commented on src/types/duration.rs:
```diff
@@ -0,0 +1,59 @@
+impl DurationWrapper {
+  #[napi]
+  pub fn new(months: i32, days: i32, ns_bigint: BigInt) -> napi::Result<Self> {
+    let ns = ns_bigint.get_i64();
+    if !ns.1 {
```
It's not clear at all what is `ns.1` and `ns.0`.

Let's aid readability by using deconstruction instead of `.1` access.
```suggestion
    let (_, ns) = ns_bigint.get_i64(); // Assuming that a two-element tuple is returned
    if ns {
```

adespawn commented on lib/types/duration.js:
```diff
@@ -85,7 +78,8 @@ Duration.prototype.toBuffer = function () {
 Duration.prototype.toString = function () {
-  let value = '';
+  let nanoseconds = bigint_to_long(this.internal.getNanoseconds());
+  let value = ''
```
In most cases, in JS it doesn't matter if semicolon is present. For the purpose of clean code it's best to use them, and I just missed this one.

wprzytula commented on src/result.rs:
```diff
@@ -171,47 +181,47 @@ impl CqlValueWrapper {
     // Is this correct?
     match self.internal.as_counter() {
       Some(r) => Ok(r.0.into()),
-      None => Err(Error::new(Status::GenericFailure, "Error")),
+      None => Err(Self::generic_error("counter")),
     }
```
Regarding the comment: what's the doubt here?

adespawn replied:
I wasn't able to find any usage of Counter type in the js driver. Probably will need to address, but maybe not in this PR

wprzytula replied:
So the doubt arises because there is no way to check the logic with the JS driver, because it's not triggered anywhere? Did I get it right?

adespawn replied:
Sort of. It's possibile this type is not used by the DS library so we will be adding it instead of updating as f.ex. duration type

wprzytula commented on src/types/duration.rs:
```diff
@@ -12,15 +12,15 @@ pub struct DurationWrapper {
-        "Invalid use: Should not happen?",
+        "Nanoseconds cannot overflow i64",
       ));
     }
```
Where does this guarantee come from? Is it some kind of runtime checks (where?) that prevents values out of `i64` range to be passed here?
In general, let's include such explanation upon each situation that we assume something and fail if it's not true.

adespawn replied:
This is more of an expectation, than a guarantee. In the JS version, the nanoseconds are kept in i64, and if there is bigger value provided, it quietly changes the value. (not even in c style: by taking modulo 2^k). So by throwing this error we change a behavior in (improper use / undefined situation) of Duration type

wprzytula replied:
So maybe `Nanoseconds must not overflow i64` would convey the message better? As it's our expectation/limitation that we impose on the input and we forbid the user to pass something different.

adespawn commented on lib/new-utils.js:
```diff
@@ -0,0 +1,25 @@
+ */
+function bigint_to_long(from){
+    return Long.fromString(from.toString());
+}
```
Your concerns are justified. I updated the conversion a bit, but it also may be nice to test it a bit more when it comes to performance, depending how much this function will be used.

wprzytula commented on lib/new-utils.js:
```diff
@@ -2,23 +2,34 @@
 const Long = require('long');
 
+const maxInt = BigInt(0x100000000);
+const minuOne = BigInt(-1);
```
minuOne -> minusOne

wprzytula commented on lib/new-utils.js:
```diff
@@ -2,23 +2,34 @@
 
 const Long = require('long');
 
+const maxInt = BigInt(0x100000000);
```
Is it really the maximum possible Int? Isn't e.g. `0x100000001` possible?

adespawn replied:
This is reversed engineered from how the Long type keeps it's values split between high and low

wprzytula replied:
OK, good. Then please explain this in a comment.

wprzytula commented on lib/new-utils.js:
```diff
@@ -2,23 +2,34 @@
+    if (lo < 0)
+        hi += minuOne
+    return Long.fromValue({ low: Number(lo), high: Number(hi), unsigned: false });
 }
```
Looks much better than strings :)

adespawn replied:
That took much longer than it looks like 😅

wprzytula replied:
That's perfectly understandable. Diffs rarely reflect the actual amount of work needed to get sth done.

adespawn commented on lib/new-utils.js:
```diff
@@ -0,0 +1,25 @@
+ */
+function bigint_to_long(from){
+    return Long.fromString(from.toString());
+}
```
Bigint is the way values unfit for js Number type are handled by napi and Long is the datastax way of handling 64 bit integers. 
Is this a clearer explanation than the current one?

wprzytula replied:
Yes, this definitely adds the needed context. Thanks.

adespawn commented on lib/types/duration.js:
```diff
@@ -37,40 +38,32 @@ const iso8601AlternateRegex = /P(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})/
-   * @type {Long}
-   */
-  this.nanoseconds = typeof nanoseconds === 'number' ? Long.fromNumber(nanoseconds) : nanoseconds;
+  if (nanoseconds instanceof Long)
+    this.internal = rust.DurationWrapper.new(months, days, long_to_bigint(nanoseconds));
+  else if (isNumber(nanoseconds))
+    this.internal = rust.DurationWrapper.new(months, days, BigInt(nanoseconds));
+  else
+    throw new TypeError(`Invalid nanosecond argument type: ${typeof(nanoseconds)}`);
```
If this constructor is used correctly (meaning nanoseconds are one of mentioned types), then yes.
If nanosecond is of a diffrent type, then my code fails immidietly, when the old code may partialy work, (ex. the duration type may be created, but will fail ((throw error)) when calling .toString() on it).
So this changes how the driver handles invalid usage. The correct usage remains the same.

adespawn commented:
There was a lot ot comments, and unless I missed some, I responded to all of them.

wprzytula commented on lib/types/duration.js:
```diff
@@ -161,6 +156,7 @@ Duration.fromBuffer = function (buffer) {
  * @returns {Duration}
  */
 Duration.fromString = function (input) {
+  // This does a lot of stuff. Is it nesesery to implement it in rust, or should we just keep it as is?
   const isNegative = input.charAt(0) === '-';
   const source = isNegative ? input.substr(1) : input;
   if (source.charAt(0) === 'P') {
```
Does the question still stand, or have we decided that we're keeping this logic in JS?

adespawn replied:
The question still stands

wprzytula commented on lib/types/result-set.js:
```diff
@@ -148,6 +148,8 @@ function getCqlObject(field) {
       return field.getCounter();
     case rust.CqlTypes.Double:
       return field.getDouble();
+    case rust.CqlTypes.Duration:
+      return field.getDuration();
     case rust.CqlTypes.Float:
       return field.getFloat();
```
A fundamental question: why were other types already present here, and Duration was not? What makes Duration different in this context?

adespawn replied:
Those types were added in branch ``types-support``

adespawn commented:
Brought the linear history to this branch!  No pesky history graphs will hide.

wprzytula commented:
But tests are failing... why?

adespawn commented:
Because of changes in simple query and types support, not all features of Client class and execute function are currently present, which fails examples.

ZuzaOsa commented on lib/new-utils.js:
```diff
@@ -0,0 +1,37 @@
+ * @param {bigint} from
+ * @returns {Long}
+ */
+function bigint_to_long(from) {
```
Shouldn't we keep the convention of the Datastax driver to write functions in camelCase?

wprzytula replied:
We could as well use a different case to clearly distinguish Datastax procedures and entities from ours. This is an open discussion.

adespawn commented:
I've changed from prototypes into class as discussed today, and cleaned up the commit history a little bit. If there is no new comments, I will merge this branch.

wprzytula commented on lib/new-utils.js:
```diff
@@ -0,0 +1,37 @@
+"use strict";
+
+const Long = require("long");
+
+// maxInt value is based on how does Long splits values between internal high and low fields.
+const maxInt = BigInt(0x100000000);
+const minusOne = BigInt(-1);
+
+/**
+ * Converts from bigint provided by napi into Long type used by the datastax library
+ * BigInt is the way napi handles values too big for js Number type,
+ * while Long is the way datastax parts of the code handle 64-bit integers.
+ * @param {bigint} from
+ * @returns {Long}
+ */
+function bigintToLong(from) {
+  let lo = from % maxInt;
+  let hi = from / maxInt;
+  if (lo < 0) hi += minusOne;
+  return Long.fromValue({ low: Number(lo), high: Number(hi), unsigned: false });
+}
```
We're using `Long` only because Datastax driver exposes it in its API, right? In the office we concluded that `BigInt` is available for us in the minimum supported NodeJS version, that's why I'm asking if Long is needed at all.

adespawn replied:
Not only because of that. Currently datastax in it's code uses Long for all internal i64 calculations, and changing that would require not insignificant amount of effort, thanks to functions like this: https://github.com/scylladb-zpp-2024-javascript-driver/scylladb-javascript-driver/blob/6cc023e07142e8ea6c8d8a7d49b52a40d3784a75/lib/types/duration.js#L84-L87
that assume the long is provided. Also the Long has different semantic for even basic operations like shown here:
https://github.com/scylladb-zpp-2024-javascript-driver/scylladb-javascript-driver/blob/6cc023e07142e8ea6c8d8a7d49b52a40d3784a75/lib/types/duration.js#L119-L120

wprzytula replied:
Understood. Then please create an issue about possible future replacement of Long with BigInt in the whole codebase, and how much effort it's going to be and why.

wprzytula commented on lib/types/duration.js:
```diff
@@ -28,173 +30,179 @@ const iso8601WeekRegex = /P(\d+)W/;
+   * Creates a new instance of {@link Duration}.
+   * @classdesc
+   * Represents a duration. A duration stores separately months, days, and seconds due to the fact that the number of
+   * days in a month varies, and a day can have 23 or 25 hours if a daylight saving is involved.
+   * @param {Number} months The number of months.
+   * @param {Number} days The number of days.
+   * @param {Number|Long} nanoseconds The number of nanoseconds.
+   * @constructor
    */
-  this.months = months;
-  /**
-   * Gets the number of days.
-   * @type {Number}
-   */
-  this.days = days;
+  constructor(months, days, nanoseconds) {
+    if (nanoseconds instanceof Long)
+      this.internal = rust.DurationWrapper.new(
+        months,
+        days,
+        longToBigint(nanoseconds),
+      );
```
So this is an API-compatible modernisation of Datastax API as suggested by @Chlor87?

adespawn replied:
Yes, it's cleaner way of declaring class which was suggested by Kuba.

wprzytula commented:
Don't forget to remove the temporary commit with new CI workflow.

adespawn commented:
I thought it would be better to remove this step later, either in types support or simple query when the normal CI step would be working again

adespawn commented:
As discussed today, I reverted this changes
